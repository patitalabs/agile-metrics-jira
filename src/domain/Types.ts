import {
  PtsCollectorConfig,
  PtsMetricItem,
} from './project-tracking-system/Types';

export interface TeamMetricsRequest {
  shouldUpdateEntries: boolean;
  config: any;
}

export interface CoreMetricsService {
  publish(entries: any, shouldReplaceEntries: boolean);
}

export interface PtsCollectorService {
  fetch(ciCollectorConfig: PtsCollectorConfig): Promise<PtsMetricItem[]>;
}
