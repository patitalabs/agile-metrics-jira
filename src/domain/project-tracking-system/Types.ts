export interface PtsMetricItem {
  id: string;
  dataType: string;
  createdAt: Date;
  key: string;
  createdBy: string;
  issueType: string;
  movedForward: number;
  movedBackward: number;
  storyPoints: number;
  assignees: string[];
  tags: string[];
  created: Date;
  finished: Date;
  leadTime: number;
  cycleTime: number;
  commentCount: number;
  jiraProject: string;
  teamName: string;
  estimateHealth: number;
  rawEstimateHealth: number;
  numberOfBugs: number;
}

export class PtsCollectorConfig {
  teamId: number;
  teamName: string;
  workFlowType: 'sprint' | 'kanban';
  since: Date;
  until?: Date;
  workFlowMap?: { [name: string]: number };
  fields?: { [name: string]: string };
  estimateConfig?: { maxTime: number; estimationValues: number[] };

  constructor({
    teamId,
    since,
    until = null,
    workFlowMap,
    fields,
    teamName,
    workFlowType,
  }) {
    this.teamId = teamId;
    this.since = new Date(since);
    this.until = until ? new Date(until) : null;
    this.workFlowMap = workFlowMap;
    this.fields = fields;
    this.teamName = teamName;
    this.workFlowType = workFlowType ? workFlowType : 'sprint';
  }

  isKanban(): boolean {
    return this.workFlowType === 'kanban';
  }

  isSprint(): boolean {
    return this.workFlowType === 'sprint';
  }
}
