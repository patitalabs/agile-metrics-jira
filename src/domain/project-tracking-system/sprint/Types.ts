export interface TaskStatistics {
  moveBackward: number;
  moveForward: number;
  movedToDev: Date;
}
