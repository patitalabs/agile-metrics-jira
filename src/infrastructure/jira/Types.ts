export class Sprint {
  id: number;
  name: string;
  isoStartDate: Date;
  isoEndDate: Date;
}

export class Task {
  key: string;
  typeName: string;
  statusName: string;
  assignee: string;
  epic: string;
  sprint?: Sprint;
  created: Date;
  projectName: string;
  teamName: string;
  createdBy: string;
  resolutionDate: Date;
  storyPoints: number;
  labels: string[];
  subtasks: Subtask[];
  histories: { [type: string]: HistoryEntry[] };
  numberOfComments: number;
  numberOfBugs: number;
}

export class Subtask {
  key: string;
  statusName: string;
}

export class HistoryItem {
  field: string;
  fromString: string;
  toString: string;
}

export class HistoryEntry {
  category: string;
  created: Date;
  items: HistoryItem[];
}

export interface JiraConfig {
  teamId: number;
  teamName: string;
  since: Date;
  until?: Date;
  workFlowMap?: { [name: string]: number };
  workFlowType: 'sprint' | 'kanban';
  fields?: { [name: string]: string };
}

export interface IssueDetails {
  subtasks: Subtask[];
  histories: { [type: string]: HistoryEntry[] };
  labels: string[];
  created: Date;
  createdBy: string;
  resolutionDate: Date;
  projectName: string;
  teamName: string;
  numberOfComments: number;
  numberOfBugs: number;
}

export interface JiraService {
  completedSprintsSince(
    teamId: number,
    since: Date,
    until: Date
  ): Promise<Sprint[]>;

  sprintData(jiraConfig: JiraConfig, sprint: Sprint): Promise<Task[]>;

  completedKanbanIssuesSince(jiraConfig: JiraConfig): Promise<Task[]>;
}

export interface JiraClient {
  getData(url: string): Promise<any>;
}

export interface JiraRepository {
  issueDetailsWithChangelog(
    jiraConfig: JiraConfig,
    issueId: string
  ): Promise<IssueDetails>;
  sprintData(jiraConfig: JiraConfig, sprint: Sprint): Promise<Task[]>;

  completedSprints(
    teamId: number,
    since: Date,
    until?: Date
  ): Promise<Sprint[]>;

  completedKanbanIssuesSince(jiraConfig: JiraConfig): Promise<Task[]>;
}
