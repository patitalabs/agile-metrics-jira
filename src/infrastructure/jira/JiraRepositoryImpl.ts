import {
  IssueDetails,
  JiraClient,
  JiraConfig,
  JiraRepository,
  Sprint,
  Task,
} from './Types';
import { Converters } from './Converters';
import { Utils } from '../../utils/Utils';
import { Logger } from '../../config/Logger';

export class JiraRepositoryImpl implements JiraRepository {
  constructor(private readonly jiraClient: JiraClient) {}

  async issueDetailsWithChangelog(
    jiraConfig: JiraConfig,
    issueId: string
  ): Promise<IssueDetails> {
    const url = `/rest/api/2/issue/${issueId}?expand=changelog`;
    const issueResponse = await this.jiraClient.getData(url);
    return Converters.toIssueDetails(jiraConfig, issueResponse);
  }

  async sprintData(jiraConfig: JiraConfig, sprint: Sprint): Promise<Task[]> {
    const url = `/rest/api/2/search?jql=Sprint=${sprint.id} AND NOT issuetype=Sub-task`;

    const { issues: issuesResponse } = await this.jiraClient.getData(url);

    const taskPromises: Promise<Task>[] = issuesResponse
      .filter((issue) => issue.fields.issuetype.name !== 'Sub-task')
      .map(async (issue) => this.issueDetails(jiraConfig, issue, sprint))
      .filter((issue) => Boolean(issue));
    return Promise.all(taskPromises);
  }

  async completedSprints(
    teamId: number,
    since: Date,
    until?: Date
  ): Promise<Sprint[]> {
    const url = `/rest/agile/1.0/board/${teamId}/sprint?state=closed`;
    const sprintDataResponse: any[] = await this.paginate(url, 'values');

    return sprintDataResponse
      .filter((sprintData) => {
        const completedDate: Date = new Date(sprintData.completeDate);
        return Utils.isDateInRange({ createdAt: completedDate, since, until });
      })
      .map((sprintData) => Converters.toSprint(sprintData));
  }

  async completedKanbanIssuesSince(jiraConfig: JiraConfig): Promise<Task[]> {
    const { since, until, teamId } = jiraConfig;

    const untilConstraint = until
      ? ` AND resolved <= "${Utils.formatDate(until)}" `
      : ``;
    const url = `/rest/agile/1.0/board/${teamId}/issue?jql=issuetype!=Sub-task AND resolved >= "${Utils.formatDate(
      since
    )}"${untilConstraint}`;
    const kanbanDataResponse: any[] = await this.paginate(url, 'issues');

    const taskPromises: Promise<Task>[] = kanbanDataResponse
      .map((issue) => this.issueDetails(jiraConfig, issue, null))
      .filter((issue) => Boolean(issue));

    return Promise.all(taskPromises);
  }

  private async issueDetails(
    jiraConfig: JiraConfig,
    issue,
    sprint: Sprint
  ): Promise<Task> {
    try {
      const issueDetailsWithChangelogResponse =
        await this.issueDetailsWithChangelog(jiraConfig, issue.key);
      return Converters.toTask(
        jiraConfig,
        issue,
        sprint,
        issueDetailsWithChangelogResponse
      );
    } catch (error) {
      Logger.error(`Skipping issue: ${issue.key}, error:${error.message}`);
    }
    return null;
  }

  private async paginate(
    targetUrl: string,
    responseTargetKey: string
  ): Promise<any[]> {
    let desiredStartAt = 0;
    const responseValues = [];
    let shouldContinue = false;
    do {
      const url = `${targetUrl}&startAt=${desiredStartAt}`;
      const responseData = await this.jiraClient.getData(url);
      const { startAt, maxResults, total } = responseData;

      let isLast = responseData.isLast;
      if (!isLast && isLast !== false) {
        isLast = true;
      }

      const values = responseData[responseTargetKey];
      shouldContinue = !isLast || total > maxResults;
      desiredStartAt += startAt + maxResults;
      responseValues.push(...values);
    } while (shouldContinue);

    return responseValues;
  }
}
