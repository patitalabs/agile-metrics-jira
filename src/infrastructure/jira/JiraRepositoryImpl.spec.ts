import * as issueDetailsExpanded from './test/issue-details-expanded-response.json';
import * as sprintFakeData from './test/sprint-data-response.json';
import * as boardIssuesFakeData from './test/board-issues-data-response.json';
import * as completedSprints from './test/completed-sprint-since-response.json';
import { JiraClient, JiraConfig, JiraRepository, Sprint } from './Types';
import { JiraRepositoryImpl } from './JiraRepositoryImpl';

describe('JiraRepositoryImpl', () => {
  const jiraClient: JiraClient = {
    getData: (url): Promise<any> => {
      let result = {};
      if (/\/*\/?expand=changelog/.test(url)) {
        result = issueDetailsExpanded;
      } else if (/\/search\?jql=Sprint.*/.test(url)) {
        result = sprintFakeData;
      } else if (/\/issue\?jql=.*/.test(url)) {
        result = boardIssuesFakeData;
      } else if (
        /\/rest\/agile\/.*\/board\/.*\/sprint\?state=closed/.test(url)
      ) {
        result = completedSprints;
      }
      return Promise.resolve(result);
    },
  };
  const jiraRepository: JiraRepository = new JiraRepositoryImpl(jiraClient);

  const jiraSprintConfig: JiraConfig = {
    teamId: 1,
    teamName: 'someTeamName',
    since: new Date('2018-12-03'),
    workFlowType: 'sprint',
  };

  it('should get sprintData', async () => {
    const teamId = 2;
    const sprint: Sprint = {
      id: teamId,
      isoEndDate: new Date('2018-12-14'),
      isoStartDate: new Date('2018-12-21'),
      name: 'some name',
    };

    const data = await jiraRepository.sprintData(jiraSprintConfig, sprint);
    expect(data).toMatchSnapshot();
  });

  it('should get kanbanData', async () => {
    const jiraKanbanConfig: JiraConfig = {
      teamId: 1,
      teamName: 'someTeamName',
      since: new Date('2018-12-13'),
      workFlowType: 'kanban',
    };

    const data = await jiraRepository.completedKanbanIssuesSince(
      jiraKanbanConfig
    );
    expect(data).toMatchSnapshot();
  });

  it('should get completedSprints', async () => {
    const teamId = 1;
    const since: Date = new Date('2018-12-03');

    const data = await jiraRepository.completedSprints(teamId, since);
    expect(data).toMatchSnapshot();
  });

  it('should get issueDetails', async () => {
    const issueId = 'issue-id';

    const data = await jiraRepository.issueDetailsWithChangelog(
      jiraSprintConfig,
      issueId
    );
    expect(data).toMatchSnapshot();
  });
});
