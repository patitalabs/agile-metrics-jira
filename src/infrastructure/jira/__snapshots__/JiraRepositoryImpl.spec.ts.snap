// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JiraRepositoryImpl should get completedSprints 1`] = `
Array [
  Object {
    "id": 171,
    "isoEndDate": 2018-12-21T10:47:29.639Z,
    "isoStartDate": 2018-12-14T11:59:08.967Z,
    "name": "17/11/14 Somename",
  },
]
`;

exports[`JiraRepositoryImpl should get issueDetails 1`] = `
Object {
  "created": 2018-11-08T14:45:49.815Z,
  "createdBy": "name",
  "histories": Object {
    "Epic Link": Array [
      Object {
        "category": "Epic Link",
        "created": 2018-11-08T14:45:49.917Z,
        "items": Array [
          Object {
            "field": "Epic Link",
            "fromString": null,
            "toString": "ORG-3417",
          },
        ],
      },
    ],
    "Fix Version": Array [
      Object {
        "category": "Fix Version",
        "created": 2018-11-09T11:49:50.537Z,
        "items": Array [
          Object {
            "field": "Fix Version",
            "fromString": null,
            "toString": "Q4 2018",
          },
        ],
      },
    ],
    "Rank": Array [
      Object {
        "category": "Rank",
        "created": 2018-11-13T11:03:02.807Z,
        "items": Array [
          Object {
            "field": "Rank",
            "fromString": "",
            "toString": "Ranked lower",
          },
        ],
      },
      Object {
        "category": "Rank",
        "created": 2018-11-09T11:44:38.932Z,
        "items": Array [
          Object {
            "field": "Rank",
            "fromString": "",
            "toString": "Ranked higher",
          },
        ],
      },
      Object {
        "category": "Rank",
        "created": 2018-11-09T08:07:34.379Z,
        "items": Array [
          Object {
            "field": "Rank",
            "fromString": "",
            "toString": "Ranked lower",
          },
        ],
      },
      Object {
        "category": "Rank",
        "created": 2018-11-09T08:07:14.391Z,
        "items": Array [
          Object {
            "field": "Rank",
            "fromString": "",
            "toString": "Ranked higher",
          },
        ],
      },
      Object {
        "category": "Rank",
        "created": 2018-11-08T14:53:42.810Z,
        "items": Array [
          Object {
            "field": "Rank",
            "fromString": "",
            "toString": "Ranked higher",
          },
        ],
      },
    ],
    "Sprint": Array [
      Object {
        "category": "Sprint",
        "created": 2018-11-13T11:03:02.659Z,
        "items": Array [
          Object {
            "field": "Sprint",
            "fromString": "",
            "toString": "13.11 TeamName",
          },
        ],
      },
      Object {
        "category": "Sprint",
        "created": 2018-11-09T11:44:38.594Z,
        "items": Array [
          Object {
            "field": "Sprint",
            "fromString": "",
            "toString": "",
          },
        ],
      },
      Object {
        "category": "Sprint",
        "created": 2018-11-09T08:07:34.158Z,
        "items": Array [
          Object {
            "field": "Sprint",
            "fromString": "",
            "toString": "",
          },
        ],
      },
      Object {
        "category": "Sprint",
        "created": 2018-11-09T08:07:14.073Z,
        "items": Array [
          Object {
            "field": "Sprint",
            "fromString": "",
            "toString": "",
          },
        ],
      },
    ],
    "Story Points": Array [
      Object {
        "category": "Story Points",
        "created": 2018-11-09T10:19:53.633Z,
        "items": Array [
          Object {
            "field": "Story Points",
            "fromString": null,
            "toString": "2",
          },
        ],
      },
    ],
    "Team": Array [
      Object {
        "category": "Team",
        "created": 2018-11-16T15:51:59.911Z,
        "items": Array [
          Object {
            "field": "Team",
            "fromString": "teamName",
            "toString": "teamName",
          },
          Object {
            "field": "status",
            "fromString": "PO Review",
            "toString": "Done",
          },
        ],
      },
    ],
    "assignee": Array [
      Object {
        "category": "assignee",
        "created": 2018-11-13T11:47:44.382Z,
        "items": Array [
          Object {
            "field": "assignee",
            "fromString": null,
            "toString": "Name",
          },
        ],
      },
    ],
    "description": Array [
      Object {
        "category": "description",
        "created": 2018-11-16T09:20:53.029Z,
        "items": Array [
          Object {
            "field": "description",
            "fromString": "previous",
            "toString": "new",
          },
        ],
      },
      Object {
        "category": "description",
        "created": 2018-11-09T11:02:53.832Z,
        "items": Array [
          Object {
            "field": "description",
            "fromString": "Refer to ORG-5566",
            "toString": "new desc",
          },
        ],
      },
    ],
    "status": Array [
      Object {
        "category": "status",
        "created": 2018-11-16T14:10:14.441Z,
        "items": Array [
          Object {
            "field": "status",
            "fromString": "In Progress",
            "toString": "PO Review",
          },
        ],
      },
      Object {
        "category": "status",
        "created": 2018-11-16T09:20:20.428Z,
        "items": Array [
          Object {
            "field": "status",
            "fromString": "PO Review",
            "toString": "In Progress",
          },
        ],
      },
      Object {
        "category": "status",
        "created": 2018-11-14T08:31:13.256Z,
        "items": Array [
          Object {
            "field": "status",
            "fromString": "In Progress",
            "toString": "PO Review",
          },
        ],
      },
      Object {
        "category": "status",
        "created": 2018-11-13T11:48:27.755Z,
        "items": Array [
          Object {
            "field": "status",
            "fromString": "Open",
            "toString": "In Progress",
          },
        ],
      },
    ],
  },
  "labels": undefined,
  "numberOfBugs": 1,
  "numberOfComments": 0,
  "projectName": "Project-Name",
  "resolutionDate": 2018-11-16T15:51:59.839Z,
  "subtasks": Array [
    Object {
      "key": "ORG-5708",
      "statusName": "Done",
    },
    Object {
      "key": "ORG-5709",
      "statusName": "Done",
    },
    Object {
      "key": "ORG-5710",
      "statusName": "Done",
    },
  ],
  "teamName": "someTeamName",
}
`;

exports[`JiraRepositoryImpl should get kanbanData 1`] = `
Array [
  Object {
    "assignee": "pter",
    "created": 2018-11-08T14:45:49.815Z,
    "createdBy": "name",
    "epic": "epic-name",
    "histories": Object {
      "Epic Link": Array [
        Object {
          "category": "Epic Link",
          "created": 2018-11-08T14:45:49.917Z,
          "items": Array [
            Object {
              "field": "Epic Link",
              "fromString": null,
              "toString": "ORG-3417",
            },
          ],
        },
      ],
      "Fix Version": Array [
        Object {
          "category": "Fix Version",
          "created": 2018-11-09T11:49:50.537Z,
          "items": Array [
            Object {
              "field": "Fix Version",
              "fromString": null,
              "toString": "Q4 2018",
            },
          ],
        },
      ],
      "Rank": Array [
        Object {
          "category": "Rank",
          "created": 2018-11-13T11:03:02.807Z,
          "items": Array [
            Object {
              "field": "Rank",
              "fromString": "",
              "toString": "Ranked lower",
            },
          ],
        },
        Object {
          "category": "Rank",
          "created": 2018-11-09T11:44:38.932Z,
          "items": Array [
            Object {
              "field": "Rank",
              "fromString": "",
              "toString": "Ranked higher",
            },
          ],
        },
        Object {
          "category": "Rank",
          "created": 2018-11-09T08:07:34.379Z,
          "items": Array [
            Object {
              "field": "Rank",
              "fromString": "",
              "toString": "Ranked lower",
            },
          ],
        },
        Object {
          "category": "Rank",
          "created": 2018-11-09T08:07:14.391Z,
          "items": Array [
            Object {
              "field": "Rank",
              "fromString": "",
              "toString": "Ranked higher",
            },
          ],
        },
        Object {
          "category": "Rank",
          "created": 2018-11-08T14:53:42.810Z,
          "items": Array [
            Object {
              "field": "Rank",
              "fromString": "",
              "toString": "Ranked higher",
            },
          ],
        },
      ],
      "Sprint": Array [
        Object {
          "category": "Sprint",
          "created": 2018-11-13T11:03:02.659Z,
          "items": Array [
            Object {
              "field": "Sprint",
              "fromString": "",
              "toString": "13.11 TeamName",
            },
          ],
        },
        Object {
          "category": "Sprint",
          "created": 2018-11-09T11:44:38.594Z,
          "items": Array [
            Object {
              "field": "Sprint",
              "fromString": "",
              "toString": "",
            },
          ],
        },
        Object {
          "category": "Sprint",
          "created": 2018-11-09T08:07:34.158Z,
          "items": Array [
            Object {
              "field": "Sprint",
              "fromString": "",
              "toString": "",
            },
          ],
        },
        Object {
          "category": "Sprint",
          "created": 2018-11-09T08:07:14.073Z,
          "items": Array [
            Object {
              "field": "Sprint",
              "fromString": "",
              "toString": "",
            },
          ],
        },
      ],
      "Story Points": Array [
        Object {
          "category": "Story Points",
          "created": 2018-11-09T10:19:53.633Z,
          "items": Array [
            Object {
              "field": "Story Points",
              "fromString": null,
              "toString": "2",
            },
          ],
        },
      ],
      "Team": Array [
        Object {
          "category": "Team",
          "created": 2018-11-16T15:51:59.911Z,
          "items": Array [
            Object {
              "field": "Team",
              "fromString": "teamName",
              "toString": "teamName",
            },
            Object {
              "field": "status",
              "fromString": "PO Review",
              "toString": "Done",
            },
          ],
        },
      ],
      "assignee": Array [
        Object {
          "category": "assignee",
          "created": 2018-11-13T11:47:44.382Z,
          "items": Array [
            Object {
              "field": "assignee",
              "fromString": null,
              "toString": "Name",
            },
          ],
        },
      ],
      "description": Array [
        Object {
          "category": "description",
          "created": 2018-11-16T09:20:53.029Z,
          "items": Array [
            Object {
              "field": "description",
              "fromString": "previous",
              "toString": "new",
            },
          ],
        },
        Object {
          "category": "description",
          "created": 2018-11-09T11:02:53.832Z,
          "items": Array [
            Object {
              "field": "description",
              "fromString": "Refer to ORG-5566",
              "toString": "new desc",
            },
          ],
        },
      ],
      "status": Array [
        Object {
          "category": "status",
          "created": 2018-11-16T14:10:14.441Z,
          "items": Array [
            Object {
              "field": "status",
              "fromString": "In Progress",
              "toString": "PO Review",
            },
          ],
        },
        Object {
          "category": "status",
          "created": 2018-11-16T09:20:20.428Z,
          "items": Array [
            Object {
              "field": "status",
              "fromString": "PO Review",
              "toString": "In Progress",
            },
          ],
        },
        Object {
          "category": "status",
          "created": 2018-11-14T08:31:13.256Z,
          "items": Array [
            Object {
              "field": "status",
              "fromString": "In Progress",
              "toString": "PO Review",
            },
          ],
        },
        Object {
          "category": "status",
          "created": 2018-11-13T11:48:27.755Z,
          "items": Array [
            Object {
              "field": "status",
              "fromString": "Open",
              "toString": "In Progress",
            },
          ],
        },
      ],
    },
    "key": "ORG-5785",
    "labels": undefined,
    "numberOfBugs": 1,
    "numberOfComments": 0,
    "projectName": "Project-Name",
    "resolutionDate": 2018-11-16T15:51:59.839Z,
    "sprint": null,
    "statusName": "Done",
    "storyPoints": 2,
    "subtasks": Array [
      Object {
        "key": "ORG-5708",
        "statusName": "Done",
      },
      Object {
        "key": "ORG-5709",
        "statusName": "Done",
      },
      Object {
        "key": "ORG-5710",
        "statusName": "Done",
      },
    ],
    "teamName": "someTeamName",
    "typeName": "Story",
  },
]
`;

exports[`JiraRepositoryImpl should get sprintData 1`] = `
Array [
  Object {
    "assignee": "pter",
    "created": 2018-11-08T14:45:49.815Z,
    "createdBy": "name",
    "epic": "epic-name",
    "histories": Object {
      "Epic Link": Array [
        Object {
          "category": "Epic Link",
          "created": 2018-11-08T14:45:49.917Z,
          "items": Array [
            Object {
              "field": "Epic Link",
              "fromString": null,
              "toString": "ORG-3417",
            },
          ],
        },
      ],
      "Fix Version": Array [
        Object {
          "category": "Fix Version",
          "created": 2018-11-09T11:49:50.537Z,
          "items": Array [
            Object {
              "field": "Fix Version",
              "fromString": null,
              "toString": "Q4 2018",
            },
          ],
        },
      ],
      "Rank": Array [
        Object {
          "category": "Rank",
          "created": 2018-11-13T11:03:02.807Z,
          "items": Array [
            Object {
              "field": "Rank",
              "fromString": "",
              "toString": "Ranked lower",
            },
          ],
        },
        Object {
          "category": "Rank",
          "created": 2018-11-09T11:44:38.932Z,
          "items": Array [
            Object {
              "field": "Rank",
              "fromString": "",
              "toString": "Ranked higher",
            },
          ],
        },
        Object {
          "category": "Rank",
          "created": 2018-11-09T08:07:34.379Z,
          "items": Array [
            Object {
              "field": "Rank",
              "fromString": "",
              "toString": "Ranked lower",
            },
          ],
        },
        Object {
          "category": "Rank",
          "created": 2018-11-09T08:07:14.391Z,
          "items": Array [
            Object {
              "field": "Rank",
              "fromString": "",
              "toString": "Ranked higher",
            },
          ],
        },
        Object {
          "category": "Rank",
          "created": 2018-11-08T14:53:42.810Z,
          "items": Array [
            Object {
              "field": "Rank",
              "fromString": "",
              "toString": "Ranked higher",
            },
          ],
        },
      ],
      "Sprint": Array [
        Object {
          "category": "Sprint",
          "created": 2018-11-13T11:03:02.659Z,
          "items": Array [
            Object {
              "field": "Sprint",
              "fromString": "",
              "toString": "13.11 TeamName",
            },
          ],
        },
        Object {
          "category": "Sprint",
          "created": 2018-11-09T11:44:38.594Z,
          "items": Array [
            Object {
              "field": "Sprint",
              "fromString": "",
              "toString": "",
            },
          ],
        },
        Object {
          "category": "Sprint",
          "created": 2018-11-09T08:07:34.158Z,
          "items": Array [
            Object {
              "field": "Sprint",
              "fromString": "",
              "toString": "",
            },
          ],
        },
        Object {
          "category": "Sprint",
          "created": 2018-11-09T08:07:14.073Z,
          "items": Array [
            Object {
              "field": "Sprint",
              "fromString": "",
              "toString": "",
            },
          ],
        },
      ],
      "Story Points": Array [
        Object {
          "category": "Story Points",
          "created": 2018-11-09T10:19:53.633Z,
          "items": Array [
            Object {
              "field": "Story Points",
              "fromString": null,
              "toString": "2",
            },
          ],
        },
      ],
      "Team": Array [
        Object {
          "category": "Team",
          "created": 2018-11-16T15:51:59.911Z,
          "items": Array [
            Object {
              "field": "Team",
              "fromString": "teamName",
              "toString": "teamName",
            },
            Object {
              "field": "status",
              "fromString": "PO Review",
              "toString": "Done",
            },
          ],
        },
      ],
      "assignee": Array [
        Object {
          "category": "assignee",
          "created": 2018-11-13T11:47:44.382Z,
          "items": Array [
            Object {
              "field": "assignee",
              "fromString": null,
              "toString": "Name",
            },
          ],
        },
      ],
      "description": Array [
        Object {
          "category": "description",
          "created": 2018-11-16T09:20:53.029Z,
          "items": Array [
            Object {
              "field": "description",
              "fromString": "previous",
              "toString": "new",
            },
          ],
        },
        Object {
          "category": "description",
          "created": 2018-11-09T11:02:53.832Z,
          "items": Array [
            Object {
              "field": "description",
              "fromString": "Refer to ORG-5566",
              "toString": "new desc",
            },
          ],
        },
      ],
      "status": Array [
        Object {
          "category": "status",
          "created": 2018-11-16T14:10:14.441Z,
          "items": Array [
            Object {
              "field": "status",
              "fromString": "In Progress",
              "toString": "PO Review",
            },
          ],
        },
        Object {
          "category": "status",
          "created": 2018-11-16T09:20:20.428Z,
          "items": Array [
            Object {
              "field": "status",
              "fromString": "PO Review",
              "toString": "In Progress",
            },
          ],
        },
        Object {
          "category": "status",
          "created": 2018-11-14T08:31:13.256Z,
          "items": Array [
            Object {
              "field": "status",
              "fromString": "In Progress",
              "toString": "PO Review",
            },
          ],
        },
        Object {
          "category": "status",
          "created": 2018-11-13T11:48:27.755Z,
          "items": Array [
            Object {
              "field": "status",
              "fromString": "Open",
              "toString": "In Progress",
            },
          ],
        },
      ],
    },
    "key": "ORG-5785",
    "labels": undefined,
    "numberOfBugs": 1,
    "numberOfComments": 0,
    "projectName": "Project-Name",
    "resolutionDate": 2018-11-16T15:51:59.839Z,
    "sprint": Object {
      "id": 2,
      "isoEndDate": 2018-12-14T00:00:00.000Z,
      "isoStartDate": 2018-12-21T00:00:00.000Z,
      "name": "some name",
    },
    "statusName": "Done",
    "storyPoints": 2,
    "subtasks": Array [
      Object {
        "key": "ORG-5708",
        "statusName": "Done",
      },
      Object {
        "key": "ORG-5709",
        "statusName": "Done",
      },
      Object {
        "key": "ORG-5710",
        "statusName": "Done",
      },
    ],
    "teamName": "someTeamName",
    "typeName": "Story",
  },
]
`;
